contracts/told_ya/src/lib.cairo

```diff
@@ -25,45 +29,144 @@ pub struct Prediction {
     }
 
     #[event]
     #[derive(Drop, starknet::Event)]
     enum Event {
-        OwnableEvent: OwnableComponent::Event
+        OwnableEvent: super::OwnableComponent::Event
     }
 
     #[constructor]
-    fn constructor(ref self: ContractState, initial_owner: ContractAddress){
+    fn constructor(ref self: ContractState, initial_owner: super::ContractAddress){
         self.ownable.initializer(initial_owner);
     }
 
     #[abi(embed_v0)]
-    impl HelloStarknetImpl of super::IHelloStarknet<ContractState> {
-        fn increase_balance(ref self: ContractState, amount: felt252) {
-            assert(amount != 0, 'Amount cannot be 0');
-            self.balance.write(self.balance.read() + amount);
+    impl ToldYaImpl of super::IToldYa<ContractState> {
+
+        fn create_event(ref self: ContractState, name: felt252, predictions_deadline: felt252, event_datetime: felt252, type_: felt252) -> Event_ {
+            // 1) Verifying caller is owner
+            self.ownable.assert_only_owner();
+
+            // 2) Hashing the new event
+            let hash_state = PoseidonTrait::new();
+            let hash_result: felt252 = hash_state.update(name).update(predictions_deadline).update(event_datetime).update(type_).finalize();            
+
+            // 3) Creating the instance of the new event
+            let new_event: Event_ = Event_ {
+                identifier: hash_result, 
+                name: name,
+                predictions_deadline: predictions_deadline,
+                event_datetime: event_datetime,
+                type_: type_
+            };
+
+            // 4) Storing the new event in Storage
+            self.events.write(new_event.identifier, new_event);
+
+            // 5) Adding the id to events_id
+            let mut events_id: Array<felt252> = self.events_id.read();
+            events_id.append(new_event.identifier);
+            self.events_id.write(events_id);
+
+            //6) Returning the new event
+            new_event
         }
 
-        fn get_balance(self: @ContractState) -> felt252 {
-            self.balance.read()
+        fn get_events(self: @ContractState) -> Array<Event_> {
+            let mut events_id: Array<felt252> = self.events_id.read();
+            let mut events = ArrayTrait::<Event_>::new();
+            while !events_id.is_empty(){
+                let event_id = events_id.pop_front().unwrap();
+                let event = self.events.read(event_id);
+                events.append(event);
+            };
+            events
         }
     }
 }
+
+
+
+// This block of code is used to store Array<felt252> in the Storage. 
+// TODO: Move this block in another file and import it in this file.
+impl StoreFelt252Array of starknet::Store<Array<felt252>> {
+    fn read(address_domain: u32, base: starknet::storage_access::StorageBaseAddress) -> starknet::SyscallResult<Array<felt252>> {
+        StoreFelt252Array::read_at_offset(address_domain, base, 0)
+    }
```

Comment:
As you wrote we'll move this part of code in another file later on, in a `utils.cairo` or smth like that yes
Did you have the opportunity to quickly test this part with our high level functions `create_events` and `get_events`  ?

Reply:
No, I need to learn first how to test a Starknet contract :) 
But yes, testing the contract functions I wrote is my next goal.